File: neet_code_all_cpp/array_hash/find_index_of_first_occurrence_string.cpp
#include <iostream>
#include <string>

class Solution {
    public:
        // AC
        int strStr(std::string haystack, std::string needle) {
            if(haystack.size() < needle.size()) {
                return -1;
            }

            for(auto i = 0; i < haystack.size() - needle.size() + 1; ++i) {
                int len_needle = 0;
                for(auto j = 0; j < needle.size(); ++j) {
                    if(haystack[i + j] == needle[j]) {
                        ++len_needle;
                    } else {
                        break;
                    }
                }
                if(len_needle == needle.size()) {
                    return i;
                }
            }

            return -1;
        }
};

int main(void) {
    using Case = std::pair<std::string, std::string>;
    Case case_1 = {"sadbutsad", "sad"};
    // => 0
    Case case_2 = {"leetcode", "leeto"};
    // => -1

    Solution s_1;
    std::cout << s_1.strStr(case_1.first, case_1.second) << std::endl;
    std::cout << s_1.strStr(case_2.first, case_2.second) << std::endl;
}

File: neet_code_all/src/array_hash/find_index_of_first_occurrence_string.rs
struct Solution {}
impl Solution {
    // AC
    pub fn str_str(haystack: String, needle: String) -> i32 {
        if haystack.len() < needle.len() {
            return -1;
        }

        let haystack: Vec<char> = haystack.chars().collect();
        let needle: Vec<char> = needle.chars().collect();

        for i in 0..(haystack.len() - needle.len() + 1) {
            let mut len_needle = 0;
            for j in 0..needle.len() {
                if haystack[i + j] == needle[j] {
                    len_needle += 1;
                } else {
                    break;
                }
            }
            if len_needle == needle.len() {
                return i as i32;
            }
        }

        -1
    }
}

fn main() {
    let case_1 = ("sadbutsad".to_string(), "sad".to_string());
    // => 0
    let case_2 = ("leetcode".to_string(), "leeto".to_string());
    // => -1

    println!(
        "case_1: {}",
        Solution::str_str(case_1.0.clone(), case_1.1.clone())
    );
    println!(
        "case_2: {}",
        Solution::str_str(case_2.0.clone(), case_2.1.clone())
    );
}
